### ISSUE TITLE
False negative when using literal type in invariant slot

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5321]
The following code should produce a type violation, but it doesn't.

```python
from typing import Literal, TypeVar

T = TypeVar("T")

def func1(a: T, b: T) -> T:
    return a

def func2() -> None:
    foo_list: list[Literal["foo"]] = ["foo"]
    x = func1(foo_list, [""])
    x.append("not foo")
    print(foo_list)

func2()
```

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by Hootner** [comment]
> The following code should produce a type violation, but it doesn't.
> 
> ``` python
> From typing import Literal, TypeVar
> 
> T = TypeVar("T")
> 
> def func1(a: T, b: T) -> T:
>     return a
> 
> def func2() -> None:
>     foo_list: list[Literal["foo"]] = ["foo"]
>     x = func1(foo_list, [""])
>     x.append("not foo")
>     print(foo_list)
> 
> func2()
> ```
The code you provided does not raise a type violation since Python typing hints are not enforced during runtime. These hints are useful for tools like type checkers, linters, or IDEs, which can check for potential type-related issues at development time.

In your example, func1 is a generic function that takes two arguments of the same type and returns the same type. Although foo_list is supposed to be of type list[Literal["foo"]], Python does not enforce type hints, so the runtime does not throw a type violation.

If you want to ensure typing safety in this code, I recommend using a static type checker like mypy to detect such issues before execution. With proper configuration, mypy should be able to catch this violation based on the type hints provided.

**Comment by erictraut** [comment]
This is included in pyright 1.1.315, which I just published. It will also be included in this week's insiders release of pylance.
